sopht_mpi/numeric/eulerian_grid_ops/stencil_ops_3d/laplacian_filter_mpi_3d.py

```diff
@@ -0,0 +1,417 @@
+            elementwise_copy_3d(field=field_buffer, rhs_field=filter_flux_buffer)
+        elementwise_saxpby_3d(
+            sum_field=scalar_field,
+            field_1=scalar_field,
+            field_1_prefac=1.0,
+            field_2=filter_flux_buffer,
+            field_2_prefac=-1.0,
+        )
+
+        # Laplacian filter in y direction
+        elementwise_copy_3d(field=field_buffer, rhs_field=scalar_field)
+        for _ in range(filter_order):
+            _laplacian_filter_3d_mpi(
+                filter_kernel=laplacian_filter_3d_y,
+                field_buffer=field_buffer,
+                filter_flux_buffer=filter_flux_buffer,
+            )
+            _clear_val_at_physical_domain_boundary_kernel_3d(field=filter_flux_buffer)
+            elementwise_copy_3d(field=field_buffer, rhs_field=filter_flux_buffer)
+        elementwise_saxpby_3d(
+            sum_field=scalar_field,
+            field_1=scalar_field,
+            field_1_prefac=1.0,
+            field_2=filter_flux_buffer,
+            field_2_prefac=-1.0,
+        )
+
+        # Laplacian filter in z direction
+        elementwise_copy_3d(field=field_buffer, rhs_field=scalar_field)
+        for _ in range(filter_order):
+            _laplacian_filter_3d_mpi(
+                filter_kernel=laplacian_filter_3d_z,
+                field_buffer=field_buffer,
+                filter_flux_buffer=filter_flux_buffer,
+            )
+            _clear_val_at_physical_domain_boundary_kernel_3d(field=filter_flux_buffer)
+            elementwise_copy_3d(field=field_buffer, rhs_field=filter_flux_buffer)
+        elementwise_saxpby_3d(
+            sum_field=scalar_field,
+            field_1=scalar_field,
+            field_1_prefac=1.0,
+            field_2=filter_flux_buffer,
+            field_2_prefac=-1.0,
+        )
+
+    scalar_field_filter_mpi_kernel_3d = None
+    if filter_type == "multiplicative":
+        scalar_field_filter_mpi_kernel_3d = (
+            scalar_field_multiplicative_filter_mpi_kernel_3d
+        )
+    elif filter_type == "convolution":
+        scalar_field_filter_mpi_kernel_3d = (
+            scalar_field_convolution_filter_mpi_kernel_3d
+        )
+    # Depending on the field type return the relevant filter implementation
+    if field_type == "scalar":
+        return scalar_field_filter_mpi_kernel_3d
+    elif field_type == "vector":
+
+        def vector_field_filter_kernel_3d(vector_field):
+            """
+            Applies laplacian filter on any vector field.
+            """
+            vector_field_filter_kernel_3d.kernel_support = (
+                gen_laplacian_filter_mpi_kernel_3d.kernel_support
+            )
+            scalar_field_filter_mpi_kernel_3d(scalar_field=vector_field[0])
+            scalar_field_filter_mpi_kernel_3d(scalar_field=vector_field[1])
+            scalar_field_filter_mpi_kernel_3d(scalar_field=vector_field[2])
+
+        return vector_field_filter_kernel_3d
```

Comment:
This part here can be more aptly coded up via the case switch mechanism, now possible via Python 3.10s `match-case` functionality.  Check out https://github.com/SophT-Team/SophT-Simulator/blob/4b00463c26b0cfff12f153b549fd302959f366cc/sopht/numeric/eulerian_grid_ops/stencil_ops_3d/laplacian_filter_3d.py#L178

Reply:
Very cool. Fixed in [0e17b87](https://github.com/fankiat/sopht-mpi/pull/148/commits/0e17b87002a91fb7978b3e815bdb78a8dfa577e4). I think there are probably other places that could use this refactoring. I will do that in the future in a different PR.

Reply:
Fixed in [0e17b87](https://github.com/fankiat/sopht-mpi/pull/148/commits/0e17b87002a91fb7978b3e815bdb78a8dfa577e4)